src/shared/Microsoft.Git.CredentialManager/Interop/Windows/WindowsCredentialManager.cs

```diff
@@ -9,119 +11,289 @@ namespace Microsoft.Git.CredentialManager.Interop.Windows
+            string targetName = credential.TargetName.TrimUntilIndexOf(TargetNameLegacyGenericPrefix);
+            if (!string.IsNullOrWhiteSpace(_namespace))
+            {
+                targetName = targetName.TrimUntilIndexOf($"{_namespace}:");
+            }
 
-                Win32Error.ThrowIfError(result, "Failed to write item to store.");
+            // If the target name matches the service name exactly then return 'match'
+            if (StringComparer.Ordinal.Equals(service, targetName))
+            {
+                return true;
             }
-            finally
+
+            // Try matching the target and service as URIs
+            if (Uri.TryCreate(service, UriKind.Absolute, out Uri serviceUri) &&
+                Uri.TryCreate(targetName, UriKind.Absolute, out Uri targetUri))
             {
-                if (w32Credential.CredentialBlob != IntPtr.Zero)
+                // Match host name
+                if (!StringComparer.OrdinalIgnoreCase.Equals(serviceUri.Host, targetUri.Host))
+                {
+                    return false;
+                }
+
+                // Match port number
+                if (!serviceUri.IsDefaultPort && serviceUri.Port == targetUri.Port)
+                {
+                    return false;
+                }
+
+                // Match path
+                if (!string.IsNullOrWhiteSpace(serviceUri.AbsolutePath) &&
+                    !StringComparer.OrdinalIgnoreCase.Equals(serviceUri.AbsolutePath, targetUri.AbsolutePath))
                 {
-                    Marshal.FreeHGlobal(w32Credential.CredentialBlob);
+                    return false;
                 }
+
+                // URLs match
+                return true;
             }
+
+            // Unable to match
+            return false;
         }
 
-        public bool Remove(string key)
+        private string CreateTargetName(string service, string account)
         {
-            int result = Win32Error.GetLastError(
-                Advapi32.CredDelete(key, CredentialType.Generic, 0)
-            );
+            var serviceUri = new Uri(service, UriKind.Absolute);
+            var sb = new StringBuilder();
 
-            switch (result)
+            if (!string.IsNullOrWhiteSpace(_namespace))
             {
-                case Win32Error.Success:
-                    return true;
+                sb.AppendFormat("{0}:", _namespace);
+            }
 
-                case Win32Error.NotFound:
-                    return false;
+            if (!string.IsNullOrWhiteSpace(serviceUri.Scheme))
+            {
+                sb.AppendFormat("{0}://", serviceUri.Scheme);
+            }
 
-                default:
-                    Win32Error.ThrowIfError(result);
-                    return false;
+            if (!string.IsNullOrWhiteSpace(account))
+            {
+                string escapedAccount = account.Replace('@', '_');
+                sb.AppendFormat("{0}@", escapedAccount);
             }
```

Comment:
This hunk isn't covered by tests. Seems important.

Reply:
Added a new test of the username escaping in the target name.